src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorResolver.cs

```diff
@@ -113,8 +114,15 @@ private static bool MatchesLookupInfo(TagHelperDescriptor descriptor, LookupInfo
                 return false;
             }
 
-            return string.IsNullOrEmpty(lookupInfo.TypeName) ||
-              string.Equals(descriptor.TypeName, lookupInfo.TypeName, StringComparison.Ordinal);
+            // We need to escape the TypePattern so we can choose to only allow specific regex.
+            var escaped = Regex.Escape(lookupInfo.TypePattern);
+
+            // We surround the escaped with ^ and $ in order ot ensure a regex match matches the entire 
+            // string. We also replace any '*' or '?' characters with regex to match appropriate content.
+            var pattern = "^" + escaped.Replace(@"\?", ".?").Replace(@"\*", ".*?") + "$";
```

Comment:
- similar comment on first replacement: shouldn't it be `"."` to match any single character?  when the only wildcard characters are `'?'` and `'*'` e.g. in Windows paths, they usually mean "any single character" and "any zero or more characters".
- separately how would Razor authors discover these wildcards?